contracts/specs/dai.spec

```diff
@@ -0,0 +1,364 @@
+
+    deny@withrevert(e, usr);
+
+    bool revert1 = ward != 1;
+    bool revert2 = e.msg.value > 0;
+
+    assert(revert1 => lastReverted, "Lack of auth did not revert");
+    assert(revert2 => lastReverted, "Sending ETH did not revert");
+    assert(lastReverted => revert1 || revert2, "Revert rules are not covering all the cases");
+}
+
+// Verify that balance behaves correctly on transfer
+rule transfer(address to, uint256 value) {
+    env e;
+
+    uint256 senderBalance = balanceOf(e.msg.sender);
+    uint256 toBalance = balanceOf(to);
+    bool senderSameAsTo = e.msg.sender == to;
+
+    require(toBalance + value <= max_uint); // Avoid evaluating the overflow case
+
+    transfer(e, to, value);
+
+    assert(!senderSameAsTo =>
+            balanceOf(e.msg.sender) == senderBalance - value &&
+            balanceOf(to) == toBalance + value,
+            "Transfer did not change balances as expected"
+    );
+
+    assert(senderSameAsTo =>
+            balanceOf(e.msg.sender) == senderBalance &&
+            senderBalance == toBalance,
+            "Transfer did not keep the balance in edge case as expected"
+    );
+}
+
+// Verify revert rules on transfer
+rule transfer_revert(address to, uint256 value) {
+    env e;
+
+    uint256 senderBalance = balanceOf(e.msg.sender);
+    uint256 toBalance = balanceOf(to);
+
+    transfer@withrevert(e, to, value);
+
+    bool revert1 = to == 0 || to == currentContract;
+    bool revert2 = senderBalance < value;
+    bool revert3 = e.msg.value > 0;
+
+    assert(revert1 => lastReverted, "Incorrect address didn't revert");
+    assert(revert2 => lastReverted, "Insufficient balance didn't revert");
+    assert(revert3 => lastReverted, "Sending ETH did not revert");
+    assert(lastReverted => revert1 || revert2 || revert3, "Revert rules are not covering all the cases");
+}
+
+// Verify that balance and allowance behave correctly on transferFrom
+rule transferFrom(address from, address to, uint256 value) {
+    env e;
+
+    uint256 fromBalance = balanceOf(from);
+    uint256 toBalance = balanceOf(to);
+    uint256 allowed = allowance(from, e.msg.sender);
+    bool deductAllowance = e.msg.sender != from && allowed != max_uint;
+    bool fromSameAsTo = from == to;
+
+    require(toBalance + value <= max_uint); // Avoid evaluating the overflow case
+
+    transferFrom(e, from, to, value);
+
+    assert(deductAllowance => allowance(from, e.msg.sender) == allowed - value, "Allowance did not decrease in value");
+    assert(!deductAllowance => allowance(from, e.msg.sender) == allowed, "Allowance did not remain the same");
+    assert(!fromSameAsTo => balanceOf(from) == fromBalance - value, "TransferFrom did not decrease the balance as expected");
+    assert(!fromSameAsTo => balanceOf(to) == toBalance + value, "TransferFrom did not increase the balance as expected");
+    assert(fromSameAsTo => balanceOf(from) == fromBalance && fromBalance == toBalance, "TransferFrom did not kept the balance as expected");
```

Comment:
Similar comment as above--the second condition doesn't add any assurance and can be removed.

Reply:
Also `kept` --> `keep` and `balance as` --> `balance the same as`

Reply:
This was done